src/main/scala/com/datastax/gatling/plugin/model/DseCqlStatements.scala

```diff
@@ -8,29 +8,35 @@ package com.datastax.gatling.plugin.model
 
 import java.nio.ByteBuffer
 
-import com.datastax.driver.core._
+import com.datastax.oss.driver.api.core.cql.{
+  BatchStatement => BatchS,
+  BatchStatementBuilder => BatchB,
+  BoundStatement => BoundS,
+  BoundStatementBuilder => BoundB,
+  SimpleStatement => SimpleS,
+  SimpleStatementBuilder => SimpleB,
+  _
+}
 import com.datastax.gatling.plugin.exceptions.DseCqlStatementException
 import com.datastax.gatling.plugin.utils.CqlPreparedStatementUtil
 import io.gatling.commons.validation._
-import io.gatling.core.session.{Session, _}
+import io.gatling.core.session._
 
 import scala.collection.JavaConverters._
-import scala.collection.mutable.ArrayBuffer
 import scala.util.{Try, Failure => TryFailure, Success => TrySuccess}
 
-
-trait DseCqlStatement extends DseStatement[Statement] {
-  def buildFromSession(session: Session): Validation[Statement]
-}
+trait DseCqlStatement[T <: Statement[T], B <: StatementBuilder[B,T]] extends DseStatement[B]
```

Comment:
Also wanted to be much more explicit about the type constraints here, including the recursive self-types and the relationship between statement and statement builder types.  This last part became more important once we settled on returning a builder from buildFromSession().

Reply:
💯 to use builders.

As for the type parameters, I find them frustrating because they propagate everywhere in the codebase, but we never end up using them anywhere: everything we call on a `B` is defined on the parent `StatementBuilder` type, and everything we do with a `T` could be done with a `Statement[_]` (`session.executeAsync`, pattern-matching when we extract the query, etc).

It all boils down to that annoying `SelfT` on `StatementBuilder`. But note that we don't use that "self-type" either: the builder is only invoked in a mutable fashion:
```scala
dseAttributes.enableTrace.foreach(builder.setTracing)
```

So I think we could simply declare the builder type like this:
```scala
trait DseCqlStatement extends DseStatement[StatementBuilder[_, Statement[_]]]
```

Then we don't need `T` and `B` anywhere else anymore. For example, `CqlRequestAction` can stay unparameterized and we would have:
```scala
private def buildStatement(builder: StatementBuilder[_, Statement[_]]): Statement[_] = {
   ... // body unchanged
}
```
We can define a type alias for `StatementBuilder[_, Statement[_]]` in a package object, to avoid repeating it everywhere.

Reply:
> `StatementBuilder[_, Statement[_]]`

Yikes, that doesn't work: "type arguments do not conform to class StatementBuilder's type parameter bounds". I forgot that Scala code does not get checked instantly in IDEA 😿 
There might not be a way around those two type parameters after all.